icalcomponent_get_duration() — calculate the duration differently for VEVENT and VTODO

dilyanpalauzov opened the pull request:
For VEVENT and VAVAILABILITY calculate the distance between DTSTART and DTEND.

For VTODO calculate the distance between DTSTART and DUE.

icalcomponent_get_dtend() — return null-time, unless called on VEVENT, VAVAILABILITY or VFREEBUSY.

winterz commented:
@dilyanpalauzov could you also add a blurb to the top-level ReleaseNotes.txt about this change?  Maybe add a test to regression.c too

@mcrha to double-check: you are ok with this patch now?

winterz commented:
nit: 2 lines in your patch are longer than 100 chars.  would be appreciated if you'd split those over more than 1 line.

mcrha commented:
> @mcrha to double-check: you are ok with this patch now?

Yes, I am. It makes the evolution-data-server tests pass without additional changes (but I'll keep the changes there, because they make sense).

ksmurchison commented on src/libical/icalcomponent.c:
```diff
@@ -1406,6 +1406,12 @@ struct icaltimetype icalcomponent_get_dtstart(icalcomponent *comp)
 {
     icalcomponent *inner = icalcomponent_get_inner(comp);
+    const icalcomponent_kind kind = icalcomponent_isa(inner);
+    if (kind != ICAL_VEVENT_COMPONENT && kind != ICAL_VAVAILABILITY_COMPONENT
```
I think a switch() statement would be cleaner (and more extendible) than an if.  I need to tweak the VPOLL support to match the current IETF draft, and will have to add it in here eventually.

ksmurchison commented on src/libical/icalcomponent.c:
```diff
@@ -1505,12 +1524,13 @@ struct icaldurationtype icalcomponent_get_duration(icalcomponent *comp)
          */
         struct icaltimetype start = icalcomponent_get_dtstart(inner);
-        struct icaltimetype end = icalcomponent_get_dtend(inner);
+        struct icaltimetype end = kind == ICAL_VTODO_COMPONENT
```
Can't we grab the end time in the new switch statement above?

ksmurchison reviewed (changes requested):
The logic looks good to me, but I have a couple of comments inline.

dilyanpalauzov commented:
I have added text to ReleaseNotes.txt, added tests to regression.c, switched to `switch()`, extended the documentation of `icalcomponent_get_duration()` and `icaltimetype icalcomponent_get_dtend()` about the implicit calculations for VEVENT.

dilyanpalauzov commented:
I added handling for the AVAILABILITY component, apart from VTODO, VEVENT, VFREEBUSY and VAVAILABILITY components.

dilyanpalauzov commented:
I wrote `AVAILABILITY`, but I meant `AVAILABLE`.

winterz commented:
I think we need this in the 3.0 branch as well.  The patch doesn't apply cleanly to 3.0, but I can cherry-pick and fix it, hopefully.
Any other final comments @ksmurchison or @mcrha ?

ksmurchison reviewed (approved):
LGTM

winterz commented on src/libical/icalcomponent.c:
```diff
@@ -1406,6 +1406,18 @@ struct icaltimetype icalcomponent_get_dtstart(icalcomponent *comp)
+        return icaltime_null_time();
+    }
+
     icalproperty *end_prop = icalcomponent_get_first_property(inner, ICAL_DTEND_PROPERTY);
```
declare `icalproperty *end_prop, *dur_prop` and `struct icaltimetype ret` above the switch else some older C compilers will barf.

winterz commented on src/libical/icalcomponent.c:
```diff
@@ -1488,12 +1504,32 @@ void icalcomponent_set_duration(icalcomponent *comp, struct icaldurationtype v)
+        return icaldurationtype_null_duration();
+    }
 
     icalproperty *dur_prop = icalcomponent_get_first_property(inner, ICAL_DURATION_PROPERTY);
```
same here. `icalproperty *dur_prop` and `struct icaldurationtype ret` must be declared above the switch

winterz reviewed (commented):
fixes needed for old C compilers

dilyanpalauzov commented:
I removed the handling of AVAILABLE, and left VAVAILABILITY.   The `icalcomponent_get_duration()` and `icalcomponentget_dtend()` functions get as parameter a VCALENDAR component and act on its first child component. AVAILABLE is not direct child of VCALENDAR.

mcrha commented:
> functions get as parameter a VCALENDAR component and act on its first child component. AVAILABLE is not direct child of VCALENDAR.

Passing a VCALENDAR component feels like a misuse of the function. You should not expect anything about the caller's intention when he/she does so. The first child component can be a VTIMEZONE or any other. I mean, I would not use the first child, when the passed in component is not one of the expected component types, just return a null-time.

dilyanpalauzov commented:
I was wrong: `icalcomponent_get_duration()` and `icalcomponet_get_dtent()` call `icalcomponent_get_inner()`.  For VCALENDAR `icalcomponent_get_inner()` returns the first child, otherwise it returns the passed parameter.

I have restored the handling of the AVAILABLE component.

dilyanpalauzov commented:
Removing `icalcomponent *inner = icalcomponent_get_inner(comp);` from `icalcomponent_get_dtend()` and others shall be handled in a different ticket.

mcrha commented:
> Removing `icalcomponent *inner = icalcomponent_get_inner(comp);` from `icalcomponent_get_dtend()` and others shall be handled in a different ticket.

I agree. I understood from your previous comment that you added that there. If it was there already, then it can be some library users depend on the current behavior. In that case it's safer to be left there, even I still believe it's a bad idea.

winterz commented:
I'll work on cherry-picking this back to the 3.0 branch.
unless there are objections?

mcrha commented:
No objection from my side. The evolution-data-server unit tests pass with it with no problem (the apply of the patch reports a lot of offset-s, but otherwise applies the code changes cleanly).
